### PR TITLE
datasources vcd_resource_list and vcd_resource_schema

### PR DESCRIPTION
**Pull request by dataclouder**
## `vcd_resource_list`

* List deployed resources by name, ID,or  HREF
* List deployed resources with import command

data source gives information about VCD, with a list of available resources, in several formats.

```hcl
data "vcd_resource_list" "list_of_orgs" {
  name          = "list_of_orgs"
  resource_type = "org"
  list_mode     = "name_id"
}

// Shows the list of organizations with the corresponding ID
output "org_list" {
  value = data.vcd_resource_list.list_of_orgs.list
}
```

## `vcd_resource_schema`

* Shows the structure of any resource in terraform-provider-vcd
```hcl
data "vcd_resource_schema" "net_struct" {
  name "net_struct"
  resource_type = "vcd_network_isolated"
}

// Shows the structure of resource network_isolated
output "net_struct" {
  value = data.vcd_resource_schema.net_struct
}
```
More examples in the docs page.
This PR resumes work started in PR #415

### CODE REVIEW

**Review by Didainius — COMMENTED** [review]
Great things with immediate use cases now:) 

A few general comments on docs:
* Looks like HCL examples don't like `//` for comments when highlighting ( at least the old engine). I think default HCL comment `#` should reender it better:
image

* Given that it is not a standard resource maybe "Example Usage 1...n" headers could have titles. Like "Example Usage 1 (List all organizations)" instead of just "Example Usage 1":
image

File: CHANGELOG.md [hunk 1]
```diff
@@ -176,6 +178,7 @@ NOTES:
   data races. Additionally GNUMakefile supports `make installrace` and `make buildrace` to build
   binary with race detection enabled. ([#453](https://github.com/vmware/terraform-provider-vcd/issues/453))
 * Add `make test-upgrade-prepare` directive ([#462](https://github.com/vmware/terraform-provider-vcd/issues/462))
+>>>>>>> master
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply CHANGELOG.md]: thanks. Removed

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,5 @@
+## 3.1.0 (Unreleased)
+
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply CHANGELOG.md:2]: done

File: vcd/datasource_vcd_resource_list.go [hunk 3]
```diff
@@ -0,0 +1,717 @@
...
+				Optional:    true,
+				Description: "The name of VDC to use, optional if defined at provider level",
+			},
+			// Parent will be needed for:
+			// * VM (parent: vApp)
+			// * catalogItem (catalog)
+			// * mediaItem (catalog)
+			// * all edge gateway objects (NAT, firewall, lb)
+			// When the parent is org or vdc, they are taken from the regular fields above
+			"parent": {
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_resource_list.go:42]: I have added some examples in the docs

File: vcd/datasource_vcd_resource_list_test.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:          func() { testAccPreCheck(t) },
+		ProviderFactories: testAccProviders,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: configText,
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(
+						"data.vcd_resource_list."+def.name, "name", def.name),
+					//resource.TestCheckResourceAttr( "data.vcd_resource_list.orgs", "list.0", "System"),
```
[the remark this answers is not in the document]
Comment (dataclouder) [reply vcd/datasource_vcd_resource_list_test.go]: removed

**Review by Didainius — APPROVED** [review]

File: vcd/datasource_vcd_resource_list.go [hunk 5]
```diff
@@ -0,0 +1,720 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The name of the parent to the resources being retrieved",
+			},
+			"name": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Unique name of the Info",
+			},
+			"resource_type": &schema.Schema{
```
Comment (vbauzysvmware) [inline vcd/datasource_vcd_resource_list.go:52]: maybe worth it to list possible values in description and validate them

File: vcd/datasource_vcd_resource_list.go [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+		}
+	}
+	return list, err
+}
+
+func externalNetworkList(d *schema.ResourceData, meta interface{}) (list []string, err error) {
+	client := meta.(*VCDClient)
+
+	if !client.VCDClient.Client.IsSysAdmin {
+		return []string{}, nil
```
Comment (vbauzysvmware) [inline vcd/datasource_vcd_resource_list.go:124]: Here I would return better error then expect user to know that he iusing none sys user

**Review by vbauzysvmware — COMMENTED** [review]
LGTM. I know we are not experts in HCL scripting, but it would be good to figure out for ourselves and for others to write in documentation(examples) for e.g. how to take out value from this data and use in next resource:
```
output: 
list_networks_hierarchy = [
  "datacloud  vdc-datacloud  net-datacloud-r",
  "datacloud  vdc-datacloud  net-datacloud-i",
  "datacloud  vdc-datacloud  net-datacloud-r2",
  "datacloud  vdc-datacloud  net-datacloud-d",
]
```

File: vcd/datasource_vcd_resource_list.go (hunk 5 above)
Comment (dataclouder) [reply vcd/datasource_vcd_resource_list.go:52]: added

File: vcd/datasource_vcd_resource_list.go (hunk 6 above)
Comment (dataclouder) [reply vcd/datasource_vcd_resource_list.go]: changed

**Review by vbauzysvmware — APPROVED** [review]

File: .changes/v3.1.0/594-features.md [hunk 7]
```diff
@@ -0,0 +1,2 @@
+**New Data source:** `vcd_resource_info` info about VCD entities [GH-594]
```
Comment (lvirbalas) [inline .changes/v3.1.0/594-features.md:1]: I believe a better name is `vcd_resource_list` and it looks you are even using it :) The old name is here and in PR name, description and maybe somewhere else too.

File: website/docs/d/resource_list.html.markdown [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+The following arguments are supported:
+
+* `org` - (Optional) The name of organization to use, optional if defined at provider level
+* `vdc` - (Optional) The name of VDC to use, optional if defined at provider level
+* `name` - (Required) An unique name to identify the data source
+* `resource_type` (Required) Which resource we want to list. Supported keywords are:
+    * `resources`  (list the resource types in the provider)
+    * `vcd_org`
+    * `vcd_external_network`
```
Comment (lvirbalas) [inline website/docs/d/resource_list.html.markdown:266]: We should add the NSX-T resource support from 3.0.0 too.
CC: @Didainius 

BTW, in general I'm concerned that this resource is manually programmed and adds a custom programming effort with each new added resource in the future. Is there any way to program it in such a way that existing resources are automatically scanned and exposed, instead?

**Review by lvirbalas — CHANGES_REQUESTED** [review]
One tiny thing and one huge question.

File: .changes/v3.1.0/594-features.md (hunk 7 above)
Comment (dataclouder) [reply .changes/v3.1.0/594-features.md]: fixed

File: website/docs/d/resource_list.html.markdown (hunk 8 above)
Comment (dataclouder) [reply website/docs/d/resource_list.html.markdown:266]: We went through this argument in PR #415, and that was the main reason for it being abandoned. 
Listing capability should be a trait of a new resource just like import.
The corresponding function for each entity should be implemented in govcd and added in this data source.

The goal of this resumed feature is to facilitate migration.

Comment (dataclouder) [reply website/docs/d/resource_list.html.markdown:266]: > We should add the NSX-T resource support from 3.0.0 too.

The NSXT VDC and external network v2 are already found with the corresponding old resources. 
I am looking at the possibility of differentiating these resources with separate lists. The fact us that even in the UI they are collected together

Comment (dataclouder) [reply website/docs/d/resource_list.html.markdown:266]: Documentation updated

File: website/docs/d/resource_list.html.markdown [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+/* 
+output:
+  "org_list" = [
+    "System  urn:vcloud:org:a93c9db9-7470-3191-8d19-a8f7eeda87f9",
+    "my-org  urn:vcloud:org:92554cc7-6222-4102-af48-364c95ed1a35",
+  ]
+*/
+```
+
+## Example 3 - List of network s - output import
```
Comment (lvirbalas) [inline website/docs/d/resource_list.html.markdown:63]: Typo in `network s`.

File: website/docs/d/resource_list.html.markdown [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+  }
+  {
+    "description" = ""
+    "dhcp_pool" = []
+    "dns1" = ""
+    "dns2" = ""
+    "dns_suffix" = ""
+    "edge_gateway" = "gw-datacloud"
+    "gateway" = "192.168.3.1"
+    "href" = "https://vcd.example.com/api/network/2cc713b1-134f-4f21-9208-79f1e4f3ee36"
```
Comment (lvirbalas) [inline website/docs/d/resource_list.html.markdown:182]: Rendering breaks at here:

Screenshot 2020-11-25 at 13 40 59

**Review by lvirbalas — CHANGES_REQUESTED** [review]

File: website/docs/d/resource_schema.html.markdown [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+Showing a structure with simple attributes only 
+
+```hcl
+data "vcd_resource_schema" "org_struct" {
+  name          = "org_struct"
+  resource_type = "vcd_org"
+}
+
+// Shows the organization attributes
+output "org_struct" {
```
Comment (lvirbalas) [inline website/docs/d/resource_schema.html.markdown:26]: Something is wrong with the computed fields. Try running the `apply` for the second time with exactly the same configuration and you'll end up with many changes:
Screenshot 2020-11-25 at 14 04 17

Comment (dataclouder) [reply website/docs/d/resource_schema.html.markdown:26]: Changed the representation from list to set. It now ignores listing order

File: website/docs/d/resource_list.html.markdown (hunk 10 above)
Comment (dataclouder) [reply website/docs/d/resource_list.html.markdown:192]: fixed

File: website/docs/d/resource_list.html.markdown (hunk 9 above)
Comment (dataclouder) [reply website/docs/d/resource_list.html.markdown]: fixed

**Review by lvirbalas — APPROVED** [review]
Looks good now and very useful!

**Pull request merged**